pycbc/inference/distributions.py

```diff
@@ -1457,13 +1457,244 @@ def from_config(cls, cp, section, variable_args):
+            self._lognorm = numpy.log(self._norm)
+
+    @property
+    def norm(self):
+        return self._norm
+
+    @property
+    def lognorm(self):
+        return self._lognorm
+
+    def rvs(self, size=1, param=None):
+        """Gives a set of random values drawn from this distribution.
+
+        Parameters
+        ----------
+        size : {1, int}
+            The number of values to generate; default is 1.
+        param : {None, string}
+            If provided, will just return values for the given parameter.
+            Otherwise, returns random values for each parameter.
+
+        Returns
+        -------
+        structured array
+            The random values in a numpy structured array. If a param was
+            specified, the array will only have an element corresponding to the
+            given parameter. Otherwise, the array will have an element for each
+            parameter in self's params.
+        """
+        if param is not None:
+            dtype = [(param, float)]
+        else:
+            dtype = [(p, float) for p in self.params]
+        arr = numpy.zeros(size, dtype=dtype)
+        for (p,_) in dtype:
+            arr[p] = numpy.random.uniform(0.0, 1.0, size=size)
+            arr[p] = self._bounds[p][1] * numpy.power(arr[p], 1.0 / self.dim)
+        return arr
+
+    def pdf(self, **kwargs):
+        """Returns the pdf at the given values. The keyword arguments must
+        contain all of parameters in self's params. Unrecognized arguments are
+        ignored. Any boundary conditions are applied to the values before the
+        pdf is evaluated.
+        """
+        return self._pdf(**self.apply_boundary_conditions(**kwargs))
+
+    def _pdf(self, **kwargs):
+        """The underlying pdf function called by `self.pdf`. This must be set
+        by any class that inherits from this class. Otherwise, a
+        `NotImplementedError` is raised.
+        """
+        for p in self._params:
+            if p not in kwargs.keys():
+                raise ValueError(
+                            'Missing parameter {} to construct pdf.'.format(p))
+        if kwargs in self:
+            pdf = self._norm * \
+                  numpy.prod([(kwargs[p])**(self.dim - 1)
+                              for p in self._params])
+            return float(pdf)
+        else:
+            return 0.0
+
+    def _logpdf(self, **kwargs):
+        """The underlying pdf function called by `self.pdf`. This must be set
+        by any class that inherits from this class. Otherwise, a
+        `NotImplementedError` is raised.
+        """
+        for p in self._params:
+            if p not in kwargs.keys():
+                raise ValueError(
+                            'Missing parameter {} to construct pdf.'.format(p))
+        if kwargs in self:
+            log_pdf = self._lognorm + \
+                      (self.dim - 1) * \
+                      sum([numpy.log(kwargs[p]) for p in self._params])
```

Comment:
Do: `numpy.log([kwargs[p] for p in self._params]).sum()` instead. It's about an order of magnitude faster for longer lists (granted, we might never have a dictionary large enough for which this makes a difference, but might as well do it since it's simple).